Implemented tls_x509_crl resource

fllaca opened the pull request:
Hi, I added here a first version of a x509 CRL resource, resolving #20 . The way of using it is using a field `certs_to_revoke`, that contains the certificates (in PEM format) that will be revoked.

Resolves #20

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+			},
+			Required:    true,
+			Description: "PEM-encoded certificates to be revoked",
+			ForceNew:    true,
```
While technically correct, it would be better, if we would update the list of certificates in the CRL, rather than create the new one every time, as then the expiry date will be updated too for example.

fllaca replied:
Yeah... but I'm quite unsure on how to handle that lifecycle 🤔 , for instance the Id of the resource is calculated as a hash of the CRL PEM content (in the same way we do for locally signed certificates and other resources in this provider). When a CRL is "updated" actually what happens is that a new version of it is created, basically a new CRL itself. I was checking https://tools.ietf.org/html/rfc5280#section-5.1 trying to find a field that we can use to identify a CRL independently of the version, but I couldn't find any. But I can miss something, do you have any idea around this?

invidian replied:
Hm, I'm not sure if I understand the problem. If you set the ID only in `Create` function and then don't touch it in `Update`, then what's the problem? I don't think there is a need to change the ID.

Aside from that, I think this is completely fine for the initial implementation. It can always be improved later :+1:

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+		"ca_key_algorithm": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Name of the algorithm used to generate the certificate's private key",
```
This description doesn't seem right.

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+	return nil
+}
+
+func UpdateCRL(d *schema.ResourceData, meta interface{}) error {
```
I think this function can be removed for now, if we don't support updates.

fllaca replied:
It was kept because of the `early_renewal_hours` field. If we make the resource non-updatable that field needs to be flagged with `ForceNew: true`. Does it make sense?

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+		}
+	}
+
+	if readyForRenewal {
```
I would move this block of code to separate function, let's call it `markForRenewal()`, so then at the beginning of this function, you could do the following:
```go
	err := endTime.UnmarshalText([]byte(endTimeStr))
	if err != nil {
          return markForRenewal(d)
        }
```

This way, you can avoid using `else`, which generally should be avoided in Go (e.g. https://medium.com/@jamousgeorges/dont-use-else-66ee163deac3).

`markForRenewal()` would be something like:
```go
func markForRenewal(d *schema.ResourceDiff) error {
  err := d.SetNew("ready_for_renewal", true)
  if err != nil {
    return err
  }

  return d.ForceNew("ready_for_renewal")
}
```

fllaca replied:
Done, love this suggestion! it reduces the cognitive complexity making it way easier to understand, thanks! Thanks also for the article reference, interesting! :)

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+			if err != nil {
+				return err
+			}
+			certificate, err := x509.ParseCertificate(block.Bytes)
```
I wonder if `parseCertificate` could be somehow refactored, so it can be used here and in the tests. I guess PEM decoding + parsing x509 could be always done together.

fllaca replied:
Done!

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	crlPem := string(pem.EncodeToMemory(&pem.Block{Type: "X509 CRL", Bytes: crlBytes}))
+	if err != nil {
```
This error check is not needed I think.

fllaca replied:
right, removed

invidian commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,89 @@
+						return fmt.Errorf("output for \"crl_pem\" is not a string")
+					}
+					if !strings.HasPrefix(got, "-----BEGIN X509 CRL----") {
+						return fmt.Errorf("key is missing cert PEM preamble")
```
```suggestion
						return fmt.Errorf("key is missing CRL PEM preamble")
```

invidian commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,89 @@
+						return fmt.Errorf("error parsing crl: %s", err)
+					}
+
+					// Verify testCertificate is in the CRL serial number list
```
```suggestion
					// Verify testCertificate is in the CRL serial number list.
```

invidian commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,89 @@
+						return fmt.Errorf("revoked certificate serial number doesn't match")
+					}
+
+					// Verify CRL signature with CA
```
```suggestion
					// Verify CRL signature with CA.
```

invidian commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,89 @@
+func certRevocationListConfig(validity uint32, earlyRenewal uint32) string {
+	return fmt.Sprintf(`
+	                locals {
+						cert_to_revoke = <<EOT
```
I think indentation here could be improved. Perhaps tabs and spaces are mixed?

fllaca replied:
should be fixed now

invidian commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+	}
+	caKey, err := parsePrivateKey(d, "ca_private_key_pem", "ca_key_algorithm")
+	if err != nil {
+		return err
```
I think it would be useful to annotate those errors using `fmt.Errorf`, so the error message is more clear for the user:
```suggestion
		return fmt.Errorf("failed to parse %q field: %w", err)
```

Also for other `return err`, which faces user.

fllaca replied:
I can do that very easily, I just followed the same error handling in other parts of the code that also do this private key parsing, for consistency with the rest of the code. Should I change it here anyway? I can also change it in the other parts a key or certificate is parsed.

invidian replied:
In my opinion, ideally all the errors should be wrapped, as far as I know it's considered a good practice (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-
gracefully) as it allows to quickly identify what exactly broke and provide better context for the user.

The comment I left was just an example on one line (pointing out every `return err` would be a lot of work :smile:), so for any new code, I'd try to follow this. I just wouldn't change the existing codebase, as this should be done in separate context (PR/issue). I hope that make sense.

fllaca replied:
Agree, done! wrapped the errors of the new code in this PR

fllaca commented on tls/provider_test.go:
```diff
@@ -96,6 +96,26 @@ UI+cIbsQJcr9i0SADwhqs7XYIcEz9jDqrdezenaXdGtXYg==
 -----END CERTIFICATE-----
 `
 
+var testCertificate = `
```
I just thought this would be closer to a real usage, should I change the test to use caCertificate and remove this?

invidian replied:
I was just wondering if there is some specific reason why separate certificate. I think it should be fine to re-use CA certificate, as for CA certificates, you can also use the CRLs, the same as for "end" certificates.

In my opinion, the maintainers have a deciding vote in such cases, but I would personally try to reuse the assets, which are already in place.

fllaca replied:
Agree, I juste removed the `testCert` variable and reused testCACert for this test

fllaca commented:
Hi @invidian , thanks a lot for hour review! I addressed most of your comments, please take a look again when you have a minute. For some other comments I had some doubts, responded with some thoughts about them.

fllaca commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,89 @@
+	"crypto/x509"
+	"encoding/pem"
+
+	//"errors"
```
removed!

fllaca commented on tls/resource_cert_revocation_list.go:
```diff
@@ -0,0 +1,193 @@
+
+func CreateCRL(d *schema.ResourceData, meta interface{}) error {
+	certsToRevoke := make([]pkix.RevokedCertificate, 0)
+	if value := d.Get("certs_to_revoke").([]interface{}); len(value) > 0 {
```
Done, also added a test for the case this field is empty

invidian commented on tls/resource_cert_revocation_list_test.go:
```diff
@@ -0,0 +1,298 @@
+EOT
+                    }
+                    output "crl_pem" {
+                        value = "${tls_x509_crl.test.crl_pem}"
```
Ditto `${}`.

fllaca commented:
Hi @invidian , sorry for the delayed response, I was really busy the last two weeks.

Thank you so much for the thoughtful feedback! Yeah... I see the points in your argumentation: as you say, with the current approach in this PR if you revoke a certificate and add it to the CRL you need to keep it forever in the terraform code and state (either hardcoded PEM or keeping its certificate resource definition). If you taint a certificate, the old version of it is not going to be in the list of serial numbers of the newly regenerated CRL. 

But on the other hand, if you taint it, from the crypto point of view it is going to be actually a completely new certificate different from the previous one you had in tf state. I'm not that sure that the Terraform resource should handle this usecase internally by maintaining an "history" of revoked certificates in the state that don't appear in the resource definition in the terraform code. Normally, what I expect from terraform is that the code, the state, and the actual real resources are three views of the same thing, and Terraform's goal is to reconcile all three. Anyway I can be wrong here, I don't know if there's any other existing provider/resource that does something similar 🤔 , it could serve as inspiration to implement this as well.

These are just some thoughts, but I don't have a strong opinion... I agree with you, it would be good having some more input feedback on this. In any case I will give it a try to implement a version that keeps the history of revoked certs, maybe in another branch, at least just to see how it looks.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-tls?pullRequest=73)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-tls?pullRequest=73)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-tls?pullRequest=73).</sub>

mhmdio commented:
any update on this @fllaca ?

swade1987 commented:
any update on this?

plsanchezfaure commented:
any update on this?

sergio-toro commented:
Hey any update on this PR? would be great to manage revocation lists with this module 🙏🏽

trebidav commented:
Would love to see this implemented! It's needed for years already!

babicamir commented:
Any updates!?
Security concerns are growing!?
Thanks

circa10a commented:
Still growing!
